pkg/server/event_handler.go

```diff
@@ -234,33 +240,57 @@ func (h *EventHandler[T, P]) HTTPHandler() http.Handler {
 // [GCS notification]: https://cloud.google.com/storage/docs/pubsub-notifications#format
 func (h *EventHandler[T, P]) Handle(ctx context.Context, m pubsub.Message) error {
 	logger := logging.FromContext(ctx)
+
+	eventBytes, processErr := h.generatePmapEventBytes(ctx, m)
+
+	attr := map[string]string{}
+
+	if processErr != nil {
+		if errors.Is(processErr, pmaperrors.ErrPubsubRetryable) {
+			return processErr
+		}
+		attr[PubSubAttrKeyNonRetryableErr] = processErr.Error()
+		logger.Errorw(processErr.Error(), "bucketId", m.Attributes["bucketId"], "objectId", m.Attributes["objectId"])
+		if err := h.failureMessenger.Send(ctx, eventBytes, attr); err != nil {
+			return fmt.Errorf("failed to send failure event downstream: %w", err)
+		}
+		return nil
+	}
+	if err := h.successMessenger.Send(ctx, eventBytes, attr); err != nil {
+		return fmt.Errorf("failed to send succuss event downstream: %w", err)
+	}
+	return nil
+}
+
+func (h *EventHandler[T, P]) generatePmapEventBytes(ctx context.Context, m pubsub.Message) ([]byte, error) {
+	// TODO(#20): The currently logic is to treat all errors that are not caused by processor as retryable.
+	// and all errors that are caused by processor as non-retryable.
+	// Therefore, we intend to keep this logic here, and do a more detailed differentiate later.
+
 	// Get the GCS object as a proto message given GCS notification information.
 	p, err := h.getGCSObjectProto(ctx, m.Attributes)
 	if err != nil {
-		return fmt.Errorf("failed to get GCS object: %w", err)
+		return nil, fmt.Errorf("%w: failed to get GCS object: %w", pmaperrors.ErrPubsubRetryable, err)
```

Comment:
Having to add such pmaperrors.ErrPubsubRetryable is weird and is easy to get ignored. I think by defining the pmap error follows https://github.com/sethvargo/go-retry/blob/a99f7cdbe5d1697463eccaf888be309b9e8d1b5f/retry.go#L24 can help solve this issue.